### ISSUE TITLE
Disk based file system 

### ISSUE THREAD

**Issue by PhilAndrew** [puniverse/javafs#3]
Does there exist a disk based file system which can work with this? I mean files and folders which store on the normal file system. I cannot find...

**Comment by pron** [comment]
I don't understand what you mean. Why implement a custom file system if what you want is the OS's file system?  In any event, the JDK comes with a Zip FS, that stores and reads files to and from zip files.

**Comment by PhilAndrew** [comment]
Hi, thanks to reply. Assume that I am trying to make an Amazon S3 based file system (not actually S3 but something similar) which works with your javafs. So what I have decided to do is to take JimFS and change it so that it can store the files on the hard disk in the file system as a first step towards that goal. That part is done but not well tested, just working.

Then if a file is read and it doesn't exist on the hard disk then I pull it from S3 to the hard disk and again then return the file it via my adapted JimFS which goes to the disk and gets the file.

So, I was thinking maybe I made life too hard for myself adapting JimFS to be a disk file system based FS, maybe I should have used something else, but ... in the end its working. There is a disk based file system in Java right, I looked and found some code, I could have changed that, but I didn't test it with JavaFS.

Do you know of any other disk based FileSystem?

**Comment by pron** [comment]
Well, there's the default file system provider. You could wrap that and combine it with [this S3 file system](https://github.com/Upplication/Amazon-S3-FileSystem-NIO2).

**Comment by btiernay** [comment]
+1 

I independently came to a similar conclusion about this combination. Adding some use cases to the README.md might be nice. I have added another one here: http://stackoverflow.com/questions/22966176/creating-a-custom-filesystem-implementation-in-java/32887126#32887126, which incidentally @PhilAndrew, shows how to customize to use and external API.

**Comment by PhilAndrew** [comment]
Thanks for the information! 

Just to share my progress, what I did was I converted  jimfs to use the disk as the storage in files and folders. This was a little over a weeks work to get it working. Then since I store files remotely, if a file is not on the file system but requested for read then it pulls the file from the remote. Then it's in the filesystem and can be returned to client via this. 

All up this was a lot of work. Perhaps there was an easier way,  but the good thing about this is that jimfs has all the features I needed such as posix attributes, file permissions, adapting it means I still retain those features. 
Also I get a good level of control over the behaviour of the file system. 

I think there are easier ways, one way I noticed is that hadoop has some filesystem plugins,  also hadoop filesystem can mount as a fuse.  Perhaps that would have been easier but I worry it wouldn't have had all the features I needed. In which case I may have been struggling to program them in.
